README.md
# Auto Mail Accounts
The Nextcloud [mail app](https://apps.nextcloud.com/apps/mail) has a built-in default configuration for mail accounts but it does not create actual mail accounts for each Nextcloud user, so they can not store alternative identitites, a sender name, attachments or collect e-mail addresses. Each user would have to enter all the mail server info (e.g. host, port, username, password) individually.

This app hooks into Nextcloud's user creation/deletion and display name/ password change and automatically creates/deletes/updates a mail account in the mail app. This creates mail accounts like if a user entered the settings manually allowing him to use all features of the mail app. This is useful if the login credentials for Nextcloud and the mail server are identical and all Nextcloud users also have a mail account.

## Features
Hooks into:
- user creation
- user deletion
- user password change
- user display name change
	- will be set as the sender name of an e-mail

## Configuration
If your Nextcloud login is identical to the e-mail address e.g. *user@example.com* and the mail server is on the same machine and uses default ports you don't need any configuration, because the default values below will be used.

If not, you can configure this app by putting the following in Nextcloud's main configuration file in *config/config.php*. These are also the default values:

	'auto_mail_accounts' => array (
                'imap_host' => 'localhost',
                'imap_port' => '143',
                'imap_ssl_mode' => 'none',
                'smtp_host' => 'localhost',
                'smtp_port' => '587',
                'smtp_ssl_mode' => 'none',
                'email_address_suffix' => ''
        ),
        
The key `auto_mail_accounts` goes into the main hierarchy level, where e.g. `trusted_domains` is.

If your users login as *user* (instead of *user@example.com*) you can set `email_address_suffix` to `@example.com` to automatically add this to the e-mail address of each user.

Valid ssl modes are `none`, `ssl` and `tls` as described in the [mail app documentation](https://github.com/nextcloud/mail/blob/master/doc/admin.md#minimal-configuration).

Note that the email field during user creation in the Nextcloud user interface is ignored and set to the username (uid).
## Security
Note that the password is stored with symmetric encryption and can be retrieved by the admin. This is the default behaviour of the mail app whose methods this app uses.

## Troubleshooting
Set `'loglevel' => '0',` in Nextcloud's *config/config.php* to enable the app's debug output and check the log file.

## Cleanup
Afaics the mail app has no mechanism to delete

- all collected e-mail addresses (i.e. addresses that are stored when you send someone an e-mail) of a user
- all references to attachments of a user
	
This means that if you delete a user, while his mail account and aliases will be deleted, his collected addresses and attachment references will remain in the db. This is not the *Auto Mail Account* app's fault.

This data can be found in the `oc_mail_attachments` and `oc_mail_collected_addresses` tables respectively.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.7] - 2020-10-13

### Fixed
- Order of XML elements in appinfo/info.xml now conforms to XSD
- removed .git files that were accidentally included

## [0.1.6] - 2020-10-07
### Added
- Compatibility with Nextcloud 20


## [0.1.5] - 2020-09-17
### Added
- Compatibility with Nextcloud 19


## [0.1.4] - 2020-03-10
### Added
- Compatibility with Nextcloud 16, 17, and 18

## [0.1.3] - 2018-12-14
### Fixed
- Fix a problem where user deletion or avatar change would fail if a user account did not have a mail account.

## [0.1.2] - 2018-12-14
### Added
- Compatibility with Nextcloud 15

## [0.1.1] - 2018-10-15
### Added
- Compatibility with Nextcloud 14

## [0.1.0] - 2018-06-07
### Added
- Initial release



appinfo/app.php
<?php
/**
 * @copyright Copyright (c) 2018 Alexey Abel <dev@abelonline.de>
 *
 * @license GNU AGPL version 3 or any later version
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

$app = new \OCA\AutoMailAccounts\AppInfo\Application();
$app->getContainer()->query('UserHooks')->register();

lib/Hooks/UserHooks.php
<?php
/**
 * @copyright Copyright (c) 2018 Alexey Abel <dev@abelonline.de>
 *
 * @license GNU AGPL version 3 or any later version
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */


namespace OCA\AutoMailAccounts\Hooks;

use OCA\Mail\Service\AccountService;
use OCA\Mail\Service\AliasesService;
use OCP\ILogger;
use OCP\IUserManager;
use OCA\Mail\Db\MailAccount;
use OCP\IUserSession;
use OCA\AutoMailAccounts\Config;

class UserHooks {

	/** @var ILogger */
	private $logger;

	/** @var Config * */
	private $config;

	/** @var IUserManager */
	private $userManager;

	/** @var IUserSession */
	private $userSession;

	/** @var AccountService */
	private $accountService;

	/** @var AliasesService */
	private $aliasesService;

	private $logContext = ['app' => 'auto_mail_accounts'];

	public function __construct(ILogger $logger, Config $config, IUserManager $userManager, IUserSession $userSession, AccountService $accountService, AliasesService $aliasesService) {
		$this->logger = $logger;
		$this->config = $config;
		$this->userManager = $userManager;
		$this->userSession = $userSession;
		$this->accountService = $accountService;
		$this->aliasesService = $aliasesService;
	}

	public function register() {
		$this->userManager->listen('\OC\User', 'postCreateUser', $this->createUserCallbackFunction());
		$this->userManager->listen('\OC\User', 'preDelete', $this->deleteUserCallbackFunction());
		$this->userManager->listen('\OC\User', 'changeUser', $this->changeUserCallbackFunction());
		$this->userManager->listen('\OC\User', 'postSetPassword', $this->changePasswordCallbackFunction());
	}

	private function createUserCallbackFunction() {
		return function (\OC\User\User $user, string $password) {

			$uid = $user->getUID();

			$account = new MailAccount();
			$account->setUserId($uid);
			$account->setName($user->getDisplayName());

			$email = $uid;
			if (!empty($this->config->getEmailAddressSuffix())) {
				$email = $email . $this->config->getEmailAddressSuffix();
			}
			$account->setEmail($email);

			$account->setInboundHost($this->config->getImapHost());
			$account->setInboundPort($this->config->getImapPort());
			$account->setInboundSslMode($this->config->getImapSslMode());
			$account->setInboundUser($email);
			$account->setInboundPassword($this->encrypt($password));

			$account->setOutboundHost($this->config->getSmtpHost());
			$account->setOutboundPort($this->config->getSmtpPort());
			$account->setOutboundSslMode($this->config->getSmtpSslMode());
			$account->setOutboundUser($email);
			$account->setOutboundPassword($this->encrypt($password));

			$this->accountService->save($account);
			$this->logger->debug("Automatically created mail account for uid " . $uid
				. " with e-mail address \"" . $email
				. "\", imap host \"" . $this->config->getImapHost()
				. "\", imap port \"" . $this->config->getImapPort()
				. "\", imap ssl mode \"" . $this->config->getImapSslMode()
				. "\", smtp host mode \"" . $this->config->getSmtpHost()
				. "\", smtp port \"" . $this->config->getSmtpPort()
				. "\", smtp ssl mode \"" . $this->config->getSmtpSslMode()
				. "\". smtp/imap user is identical to uid."
				. " e-mail address suffix was \"". $this->config->getEmailAddressSuffix(). "\"."
				, $this->logContext);
		};
	}

	private function deleteUserCallbackFunction() {
		return function (\OC\User\User $user) {
			$userUid = $user->getUID();
			$userAccounts = $this->accountService->findByUserId($user->getUID());
			// A user can have multiple configured mail accounts, so we have to
			// delete all of them. This is why findByUserId() returns an array
			// and not a single element.
			foreach ($userAccounts as $userAccount) {
				// each account can have multiple aliases
				$thisAccountsAliases = $this->aliasesService->findAll($userAccount->getId()
					, $userUid);
				foreach ($thisAccountsAliases as $alias) {
					$this->aliasesService->delete($alias->getId(), $userUid);
					$this->logger->debug("Automatically deleted alias \""
						. $alias->getName() . "\" of mail account \"" . $userAccount->getName()
						. "\".", $this->logContext);
				}
				// After deleting all aliases delete the actual mail account.
				$this->accountService->delete($userUid, $userAccount->getId());
				$this->logger->debug("Automatically deleted mail account "
					. $userAccount->getName(). "<".$userAccount->getEMailAddress()
					."> before deleting user " . $userUid.".", $this->logContext);
			}
			//TODO: delete attachments, mail app does not support deletion yet
			//TODO: delete collected e-mails, mail app does not support deletion yet
		};
	}

	private function changeUserCallbackFunction() {
		return function (\OC\User\User $user, string $feature, string $value) {
			if ($feature === 'displayName') {
				// The assumption is made that the automatically created mail
				// account is the first one, because it was automatically created
				// immediately after the creation of the user.
				$firstAccountsMailAccount = $this->getUsersFirstMailAccount($user);
				$firstAccountsMailAccount->setName($value);
				$this->accountService->save($firstAccountsMailAccount);
				$this->logger->debug("Automatically changed sender name for mail account"
					. " of uid " . $user->getUID() . " to \"" . $value . "\".", $this->logContext);
			}
		};
	}

	private function changePasswordCallbackFunction() {
		return function (\OC\User\User $user, string $password) {
			// The assumption is made that the automatically created mail
			// account is the first one, because it was automatically created
			// immediately after the creation of the user.
			$firstAccountsMailAccount = $this->getUsersFirstMailAccount($user);
			$firstAccountsMailAccount->setInboundPassword($this->encrypt($password));
			$firstAccountsMailAccount->setOutboundPassword($this->encrypt($password));
			$this->accountService->save($firstAccountsMailAccount);
			$this->logger->debug("Automatically changed password for mail account of uid "
				. $user->getUID() . ".", $this->logContext);
		};
	}

	private function encrypt($plaintext) {
		$crypto = \OC::$server->getCrypto();
		return $crypto->encrypt($plaintext);
	}

	private function getUsersFirstMailAccount(\OC\User\User $user) {
		$accounts = $this->accountService->findByUserId($user->getUID());
		uksort($accounts, function ($a, $b) {
			return ($a['account']['id'] < $b['account']['id']) ? -1 : 1;
		});

		return $accounts[0]->getMailAccount();
	}
}


lib/AppInfo/Application.php
<?php
/**
 * @copyright Copyright (c) 2018 Alexey Abel <dev@abelonline.de>
 *
 * @license GNU AGPL version 3 or any later version
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

namespace OCA\AutoMailAccounts\AppInfo;

use \OCP\AppFramework\App;
use \OCA\AutoMailAccounts\Hooks\UserHooks;

class Application extends App {

	public function __construct(array $urlParams = array()) {
		parent::__construct('auto_mail_accounts', $urlParams);

		$container = $this->getContainer();

		$container->registerService('UserHooks', function($c) {
			return new UserHooks(
				$c->query('ServerContainer')->getLogger(),
				$c->query('OCA\AutoMailAccounts\Config'),
				$c->query('ServerContainer')->getUserManager(),
				$c->query('ServerContainer')->getUserSession(),
				$c->query('OCA\Mail\Service\AccountService'),
				$c->query('OCA\Mail\Service\AliasesService')
			);
		});
	}
}

lib/Config.php
<?php
/**
 * @copyright Copyright (c) 2018 Alexey Abel <dev@abelonline.de>
 *
 * @license GNU AGPL version 3 or any later version
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

namespace OCA\AutoMailAccounts;

use OCP\ILogger;
use \OCP\IConfig;

class Config {

	const DEFAULT_IMAP_HOST = 'localhost';
	const DEFAULT_IMAP_PORT = '143';
	const DEFAULT_IMAP_SSL_MODE = 'none';
	const DEFAULT_SMTP_HOST = 'localhost';
	const DEFAULT_SMTP_PORT = '587';
	const DEFAULT_SMTP_SSL_MODE = 'none';
	const DEFAULT_EMAIL_ADDRESS_SUFFIX = '';

	const CONFIG_KEY = 'auto_mail_accounts';
	const CONFIG_KEY_IMAP_HOST = 'imap_host';
	const CONFIG_KEY_IMAP_PORT = 'imap_port';
	const CONFIG_KEY_IMAP_SSL_MODE = 'imap_ssl_mode';
	const CONFIG_KEY_SMTP_HOST = 'smtp_host';
	const CONFIG_KEY_SMTP_PORT = 'smtp_port';
	const CONFIG_KEY_SMTP_SSL_MODE = 'smtp_ssl_mode';
	const CONFIG_KEY_EMAIL_ADDRESSS_SUFFIX = 'email_address_suffix';
	
	private $logger;
	private $appConfiguration;
	private $logContext = ['app' => 'auto_mail_accounts'];

	public function __construct(ILogger $logger, IConfig $nextCloudConfiguration) {
		$this->logger = $logger;
		$this->appConfiguration = $nextCloudConfiguration->getSystemValue(self::CONFIG_KEY);
	}

	public function getImapHost() {
		return $this->getConfigValueOrDefaultValue(self::CONFIG_KEY_IMAP_HOST
			,self::DEFAULT_IMAP_HOST);
	}

	public function getImapPort() {
		return $this->getConfigValueOrDefaultValue(self::CONFIG_KEY_IMAP_PORT
			,self::DEFAULT_IMAP_PORT);
	}


	public function getImapSslMode() {
		return $this->getConfigValueOrDefaultValue(self::CONFIG_KEY_IMAP_SSL_MODE
			,self::DEFAULT_IMAP_SSL_MODE);
	}

	public function getSmtpHost() {
		return $this->getConfigValueOrDefaultValue(self::CONFIG_KEY_SMTP_HOST
			,self::DEFAULT_SMTP_HOST);
	}

	public function getSmtpPort() {
		return $this->getConfigValueOrDefaultValue(self::CONFIG_KEY_SMTP_PORT
			,self::DEFAULT_SMTP_PORT);
	}


	public function getSmtpSslMode() {
		return $this->getConfigValueOrDefaultValue(self::CONFIG_KEY_SMTP_SSL_MODE
			,self::DEFAULT_SMTP_SSL_MODE);
	}


	public function getEmailAddressSuffix() {
		return $this->getConfigValueOrDefaultValue(self::CONFIG_KEY_EMAIL_ADDRESSS_SUFFIX
			,self::DEFAULT_EMAIL_ADDRESS_SUFFIX);
	}

	/**
	 * Tries to read a config value and if it is set returns its value,
	 * otherwise returns provided value. Also logs a debug message that default
	 * value was used.
	 * @param $configKey string key name of configuration parameter
	 * @param $defaultValue string default parameter that will be returned if
	 * config key is not set
	 * @return string value of config key or provided default value
	 */
	private function getConfigValueOrDefaultValue($configKey, $defaultValue) {
		if (empty($this->appConfiguration) || empty($this->appConfiguration[$configKey])) {
			$this->logger->debug('The config key ' . $configKey
				. ' is not set, defaulting to ' . $defaultValue . '.'
				, $this->logContext);
			return $defaultValue;
		} else {
			return $this->appConfiguration[$configKey];
		}
	}
}